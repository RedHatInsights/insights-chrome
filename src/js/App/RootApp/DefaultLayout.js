import React, { memo, useEffect, useRef } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import GlobalFilter from '../GlobalFilter/GlobalFilter';
import { useScalprum } from '@scalprum/react-core';
import { Page, PageHeader, PageSidebar } from '@patternfly/react-core';
import { useLocation } from 'react-router-dom';
import { Header, HeaderTools } from '../Header/Header';
import Cookie from 'js-cookie';
import isEqual from 'lodash/isEqual';
import { onToggle } from '../../redux/actions';
import Routes from '../Routes';
import useOuiaTags from '../../utils/useOuiaTags';

import '../Sidenav/Navigation/Navigation.scss';
import './DefaultLayout.scss';
import { CROSS_ACCESS_ACCOUNT_NUMBER } from '../../consts';

const ShieldedRoot = memo(
  ({ hideNav, insightsContentRef, isGlobalFilterEnabled, initialized, Sidebar }) => {
    const dispatch = useDispatch();
    useEffect(() => {
      const navToggleElement = document.querySelector('button#nav-toggle');
      if (navToggleElement) {
        navToggleElement.onclick = () => dispatch(onToggle());
      }
    }, []);

    if (!initialized) {
      return null;
    }

    const selectedAccountNumber = Cookie.get(CROSS_ACCESS_ACCOUNT_NUMBER);
    const hasBanner = false; // Update this later when we use feature flags

    return (
      <Page
        isManagedSidebar={!hideNav}
        className={classnames({ 'ins-c-page__hasBanner': hasBanner, 'ins-c-page__account-banner': selectedAccountNumber })}
        header={<PageHeader logoComponent="div" logo={<Header />} showNavToggle={!hideNav} headerTools={<HeaderTools />} />}
        sidebar={hideNav ? undefined : <PageSidebar id="ins-c-sidebar" nav={Sidebar} />}
      >
        <div ref={insightsContentRef} className={classnames('ins-c-render', { 'ins-m-full--height': !isGlobalFilterEnabled })}>
          {isGlobalFilterEnabled && <GlobalFilter />}
          {selectedAccountNumber && <div className="ins-c-viewing-as">Viewing as Account {selectedAccountNumber}</div>}
          <Routes insightsContentRef={insightsContentRef} />
          <main className="pf-c-page__main" id="no-access"></main>
        </div>
      </Page>
    );
  },
  (prevProps, nextProps) => isEqual(prevProps, nextProps)
);

ShieldedRoot.propTypes = {
  hideNav: PropTypes.bool,
  insightsContentRef: PropTypes.object.isRequired,
  isGlobalFilterEnabled: PropTypes.bool.isRequired,
  initialized: PropTypes.bool,
  Sidebar: PropTypes.element,
};
ShieldedRoot.defaultProps = {
  useLandingNav: false,
  hideNav: false,
  isGlobalFilterEnabled: false,
  initialized: false,
};
ShieldedRoot.displayName = 'ShieldedRoot';

const RootApp = ({ globalFilterHidden, Sidebar }) => {
  const ouiaTags = useOuiaTags();
  const initialized = useScalprum(({ initialized }) => initialized);
  const { pathname } = useLocation();
  const hideNav = useSelector(({ chrome: { user } }) => !user || !Sidebar);
  const activeLocation = pathname.split('/')[1];
  /**
   * Using the chrome landing flag is not going to work because the appId is initialized inside the app.
   * We need the information before anything is rendered to determine if we use root module or render landing page.
   * This will be replaced once we can use react router for all pages. Landing page will have its own route.
   */
  const isLanding = pathname === '/';

  const isGlobalFilterEnabled =
    !isLanding && ((!globalFilterHidden && activeLocation === 'insights') || Boolean(localStorage.getItem('chrome:experimental:global-filter')));
  const insightsContentRef = useRef(null);

  return (
    <div id="chrome-app-render-root" className="pf-c-drawer__content" {...ouiaTags}>
      <ShieldedRoot
        isGlobalFilterEnabled={isGlobalFilterEnabled}
        hideNav={hideNav}
        insightsContentRef={insightsContentRef}
        initialized={initialized}
        Sidebar={Sidebar}
      />
    </div>
  );
};

RootApp.propTypes = {
  activeApp: PropTypes.string,
  pageAction: PropTypes.string,
  pageObjectId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  globalFilterHidden: PropTypes.bool,
  config: PropTypes.any,
  Sidebar: PropTypes.element,
};

export default RootApp;
