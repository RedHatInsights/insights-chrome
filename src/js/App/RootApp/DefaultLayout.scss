.ins-c-viewing-as {
  left: 0;
  top: 76px;
  width: 100%;
  padding: 2px 0 2px 20px;
  font-size: 10px;
  background-color: var(--pf-global--palette--blue-200);
  color: var(--pf-global--Color--100);
  opacity: 1;
  grid-area: banner;
}

.pf-c-page.ins-c-page__account-banner {
  grid-template-areas:
    "header header"
    "banner banner"
    "nav main";
  grid-template-rows: max-content max-content 1fr;
}

@media screen and (max-width: 1200px) {
  .pf-c-page.ins-c-page__account-banner {
    grid-template-areas:
      "header header"
      "main main";
  }
}
