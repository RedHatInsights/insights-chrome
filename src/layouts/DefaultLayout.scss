.chr-viewing-as {
  left: 0;
  top: 76px;
  width: 100%;
  padding: 2px 0 2px 20px;
  font-size: 10px;
  background-color: var(--pf-v5-global--palette--blue-200);
  color: var(--pf-v5-global--Color--100);
  opacity: 1;
  grid-area: banner;
}

.pf-v6-c-page.chr-c-page__account-banner {
  grid-template-areas:
    "header header"
    "banner banner"
    "nav main";
  grid-template-rows: max-content max-content 1fr;
}

@media screen and (max-width: 1200px) {
  .pf-v6-c-page.chr-c-page__account-banner {
    grid-template-areas:
      "header header"
      "main main";
  }
}

.chr-scope__default-layout {
  flex-grow: 1;
  display: inherit;
  flex-direction: inherit;
}

.pf-v6-c-page__drawer {
  .pf-v6-c-drawer__content {
    background-color: transparent;
  }
}

