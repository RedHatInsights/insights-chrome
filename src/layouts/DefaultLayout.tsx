import React, { memo, useContext, useEffect, useRef, useState } from 'react';
import { useAtomValue } from 'jotai';
import classnames from 'classnames';
import GlobalFilter from '../components/GlobalFilter/GlobalFilter';
import { useScalprum } from '@scalprum/react-core';
import { Masthead } from '@patternfly/react-core/dist/dynamic/components/Masthead';
import { Page } from '@patternfly/react-core/dist/dynamic/components/Page';
import { PageSidebar } from '@patternfly/react-core/dist/dynamic/components/Page';
import { PageSidebarBody } from '@patternfly/react-core/dist/dynamic/components/Page';
import { Header } from '../components/Header/Header';
import Cookie from 'js-cookie';
import isEqual from 'lodash/isEqual';
import ChromeRoutes from '../components/Routes/Routes';
import useOuiaTags from '../utils/useOuiaTags';
import RedirectBanner from '../components/Stratosphere/RedirectBanner';

import { useIntl } from 'react-intl';
import messages from '../locales/Messages';
import { CROSS_ACCESS_ACCOUNT_NUMBER } from '../utils/consts';

import DrawerPanel from '../components/NotificationsDrawer/DrawerPanelContent';

import '../components/Navigation/Navigation.scss';
import './DefaultLayout.scss';
import useNavigation from '../utils/useNavigation';
import { NavigationProps } from '../components/Navigation';
import { getUrl } from '../hooks/useBundle';
import { useFlag } from '@unleash/proxy-client-react';
import ChromeAuthContext from '../auth/ChromeAuthContext';
import VirtualAssistant from '../components/Routes/VirtualAssistant';
import { notificationDrawerExpandedAtom } from '../state/atoms/notificationDrawerAtom';
import { ITLess } from '../utils/common';

type ShieldedRootProps = {
  hideNav?: boolean;
  initialized?: boolean;
  Sidebar?: React.FC<NavigationProps>;
  Footer?: React.ReactNode;
};

type DefaultLayoutProps = {
  hasBanner: boolean;
  selectedAccountNumber?: string;
  hideNav: boolean;
  isNavOpen: boolean;
  setIsNavOpen: React.Dispatch<React.SetStateAction<boolean>>;
  Sidebar?: React.FC<NavigationProps>;
  Footer?: React.ReactNode;
};

const DefaultLayout: React.FC<DefaultLayoutProps> = ({ hasBanner, selectedAccountNumber, hideNav, isNavOpen, setIsNavOpen, Sidebar, Footer }) => {
  const intl = useIntl();
  const { loaded, schema, noNav } = useNavigation();
  const isNotificationsDrawerExpanded = useAtomValue(notificationDrawerExpandedAtom);
  const drawerPanelRef = useRef<HTMLDivElement>();
  const focusDrawer = () => {
    const tabbableElement = drawerPanelRef.current?.querySelector('a, button') as HTMLAnchorElement | HTMLButtonElement;
    tabbableElement.focus();
  };
  const isNotificationsEnabled = useFlag('platform.chrome.notifications-drawer');
  return (
    <Page
      className={
        (classnames('chr-c-page', { 'chr-c-page__hasBanner': hasBanner, 'chr-c-page__account-banner': selectedAccountNumber }),
        'pf-c-page') /** we have to add the legacy styling to allow v4 page layout sub components to be able to inherit legacy styling */
      }
      onPageResize={null} // required to disable PF resize observer that causes re-rendring issue
      masthead={
        <Masthead className="chr-c-masthead pf-v6-u-p-0" display={{ sm: 'stack', '2xl': 'inline' }}>
          <Header
            breadcrumbsProps={{
              isNavOpen,
              setIsNavOpen,
              hideNav,
            }}
          />
        </Masthead>
      }
      {...(isNotificationsEnabled && {
        onNotificationDrawerExpand: focusDrawer,
        notificationDrawer: <DrawerPanel ref={drawerPanelRef} />,
        isNotificationDrawerExpanded: isNotificationsDrawerExpanded,
      })}
      sidebar={
        (noNav || hideNav) && Sidebar
          ? undefined
          : Sidebar && (
              <PageSidebar isSidebarOpen={isNavOpen} id="chr-c-sidebar">
                <PageSidebarBody>
                  <Sidebar schema={schema} loaded={loaded} />
                </PageSidebarBody>
              </PageSidebar>
            )
      }
    >
      <div className={classnames('chr-render')}>
        <GlobalFilter key={getUrl('bundle')} />
        {selectedAccountNumber && (
          <div className="chr-viewing-as sentry-mask data-hj-suppress">
            {intl.formatMessage(messages.viewingAsAccount, { selectedAccountNumber })}
          </div>
        )}
        <RedirectBanner />
        {ITLess() ? null : <VirtualAssistant />}
        <ChromeRoutes routesProps={{ scopeClass: 'chr-scope__default-layout' }} />
        {Footer}
      </div>
    </Page>
  );
};

const ShieldedRoot = memo(
  ({ hideNav = false, initialized = false, Sidebar, Footer }: ShieldedRootProps) => {
    const [isMobileView, setIsMobileView] = useState(window.document.body.clientWidth < 1200);
    const [isNavOpen, setIsNavOpen] = useState(!isMobileView);
    /**
     * Required for event listener to access the variables
     */
    const mutableStateRef = useRef({
      isMobileView,
    });
    function navReziseListener() {
      const internalMobile = window.document.body.clientWidth < 1200;
      const { isMobileView } = mutableStateRef.current;
      if (!isMobileView && internalMobile) {
        setIsMobileView(true);
        setIsNavOpen(false);
        mutableStateRef.current = {
          isMobileView: true,
        };
      } else if (isMobileView && !internalMobile) {
        setIsMobileView(false);
        setIsNavOpen(true);
        mutableStateRef.current = {
          isMobileView: false,
        };
      }
    }

    useEffect(() => {
      window.addEventListener('resize', navReziseListener);
      return () => {
        window.removeEventListener('resize', navReziseListener);
      };
    }, []);

    if (!initialized) {
      return null;
    }

    const selectedAccountNumber = Cookie.get(CROSS_ACCESS_ACCOUNT_NUMBER);
    const hasBanner = false; // Update this later when we use feature flags

    return (
      <DefaultLayout
        setIsNavOpen={setIsNavOpen}
        hideNav={hideNav}
        isNavOpen={isNavOpen}
        hasBanner={hasBanner}
        selectedAccountNumber={selectedAccountNumber}
        Sidebar={Sidebar}
        Footer={Footer}
      />
    );
  },
  (prevProps, nextProps) => isEqual(prevProps, nextProps)
);

ShieldedRoot.displayName = 'ShieldedRoot';

export type RootAppProps = {
  Sidebar?: React.FC<NavigationProps>;
  Footer?: React.ReactNode;
};

const DefaultLayoutRoot = ({ Sidebar, Footer }: RootAppProps) => {
  const ouiaTags = useOuiaTags();
  const initialized = useScalprum(({ initialized }) => initialized);
  const { user } = useContext(ChromeAuthContext);
  const hideNav = !user || !Sidebar;

  return (
    <div id="chrome-app-render-root" {...ouiaTags}>
      <ShieldedRoot hideNav={hideNav} initialized={initialized} Sidebar={Sidebar} Footer={Footer} />
    </div>
  );
};

export default DefaultLayoutRoot;
