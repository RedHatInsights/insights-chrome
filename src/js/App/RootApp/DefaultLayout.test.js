import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import DefaultLayout from './DefaultLayout';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

describe('DefaultLayout', () => {
  let initialState;
  let mockStore;
  let config;

  beforeEach(() => {
    config = {
      foo: {
        manifestLocation: '/bar',
        appName: 'foo',
      },
    };
    mockStore = configureStore();
    initialState = {
      chrome: {
        activeApp: 'some-app',
        activeLocation: 'some-location',
        appId: 'app-id',
        navigation: {
          '/': {
            navItems: [],
          },
          insights: {
            navItems: [],
          },
        },
      },
      globalFilter: {
        tags: {},
        sid: {},
        workloads: {},
      },
    };
  });

  it('should render correctly - no data', async () => {
    const store = mockStore({ chrome: {} });
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <DefaultLayout config={config} />
        </MemoryRouter>
      </Provider>
    );
    expect(container.querySelector('.pf-c-drawer__content')).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/some-location/app-id']}>
          <DefaultLayout config={config} />
        </MemoryRouter>
      </Provider>
    );
    expect(container.querySelector('.pf-c-drawer__content')).toMatchSnapshot();
  });

  it('should render correctly with pageAction', () => {
    const store = mockStore({
      chrome: {
        ...initialState.chrome,
        pageAction: 'some-action',
      },
      globalFilter: {},
    });
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/some-location/app-id']}>
          <DefaultLayout config={config} />
        </MemoryRouter>
      </Provider>
    );
    expect(container.querySelector('.pf-c-drawer__content')).toMatchSnapshot();
  });

  it('should render correctly with pageObjectId', () => {
    const store = mockStore({
      chrome: {
        ...initialState.chrome,
        pageObjectId: 'some-object-id',
      },
    });
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/some-location/app-id']}>
          <DefaultLayout config={config} />
        </MemoryRouter>
      </Provider>
    );
    expect(container.querySelector('.pf-c-drawer__content')).toMatchSnapshot();
  });

  it('should render correctly with pageObjectId and pageAction', () => {
    const store = mockStore({
      chrome: {
        ...initialState.chrome,
        pageAction: 'some-action',
        pageObjectId: 'some-object-id',
      },
    });
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/some-location/app-id']}>
          <DefaultLayout config={config} />
        </MemoryRouter>
      </Provider>
    );
    expect(container.querySelector('.pf-c-drawer__content')).toMatchSnapshot();
  });
});
