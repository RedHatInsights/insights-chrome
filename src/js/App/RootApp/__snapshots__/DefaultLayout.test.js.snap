// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultLayout should render correctly - no data 1`] = `
<div
  class="pf-c-drawer__content"
  data-ouia-bundle=""
  data-ouia-safe="true"
  id="chrome-app-render-root"
  landing="true"
/>
`;

exports[`DefaultLayout should render correctly 1`] = `
<div
  class="pf-c-drawer__content"
  data-ouia-app-id="app-id"
  data-ouia-bundle="some-location"
  data-ouia-safe="true"
  data-ouia-subnav="some-app"
  id="chrome-app-render-root"
/>
`;

exports[`DefaultLayout should render correctly with pageAction 1`] = `
<div
  class="pf-c-drawer__content"
  data-ouia-app-id="app-id"
  data-ouia-bundle="some-location"
  data-ouia-page-type="some-action"
  data-ouia-safe="true"
  data-ouia-subnav="some-app"
  id="chrome-app-render-root"
/>
`;

exports[`DefaultLayout should render correctly with pageObjectId 1`] = `
<div
  class="pf-c-drawer__content"
  data-ouia-app-id="app-id"
  data-ouia-bundle="some-location"
  data-ouia-page-object-id="some-object-id"
  data-ouia-safe="true"
  data-ouia-subnav="some-app"
  id="chrome-app-render-root"
/>
`;

exports[`DefaultLayout should render correctly with pageObjectId and pageAction 1`] = `
<div
  class="pf-c-drawer__content"
  data-ouia-app-id="app-id"
  data-ouia-bundle="some-location"
  data-ouia-page-object-id="some-object-id"
  data-ouia-page-type="some-action"
  data-ouia-safe="true"
  data-ouia-subnav="some-app"
  id="chrome-app-render-root"
/>
`;
